admin/release.py

```diff
@@ -303,3 +329,176 @@ def publish_docs_main(args, base_path, top_level):
         sys.stderr.write("%s: Can't publish non-tagged version to production."
                          % (base_path.basename(),))
         raise SystemExit(1)
+
+
+class UploadOptions(Options):
+    """
+    Options for uploading packages.
+    """
+    optParameters = [
+        ["target", None, ARCHIVE_BUCKET,
+         "The bucket to upload packages to."],
+        ["build-server", None,
+         b'http://build.clusterhq.com',
+         "The URL of the build-server."],
+    ]
+
+    def parseArgs(self, version):
+        self['version'] = version
+
+
+FLOCKER_PACKAGES = [
+    b'clusterhq-python-flocker',
+    b'clusterhq-flocker-cli',
+    b'clusterhq-flocker-node',
+]
+
+
+@do
+def update_repo(rpm_directory, target_bucket, target_key, source_repo,
+                packages):
+    """
+    Update ``target_bucket`` yum repository with ``packages`` from
+    ``source_repo`` repository.
+
+    :param FilePath rpm_directory: Temporary directory to download
+        repository to.
+    :param bytes target_bucket: S3 bucket to upload repository to.
+    :param bytes target_key: Path within S3 bucket to upload repository to.
+    :param bytes source_repo: Repository on the build server to get packages
+        from.
+    :param list packages: List of bytes, each specifying the name of a package
+        to upload to the repository.
+    """
+    rpm_directory.createDirectory()
+
+    yield Effect(DownloadS3KeyRecursively(
+        source_bucket=target_bucket,
+        source_prefix=target_key,
+        target_path=rpm_directory,
+        filter_extensions=('.rpm',)))
+
+    yield Effect(DownloadPackagesFromRepository(
+        source_repo=source_repo,
+        target_path=rpm_directory,
+        packages=packages,
+        ))
+
+    downloaded_packages = yield Effect(ListPackages(
```

Comment:
This will get all the packages, both the ones that already exists, as well as the new ones we need to add.

Since we do know the version of the packages we are downloading,
- we could just download them directly, rather than have to use yum
- we can easily generate the list of download packages.

That would also alleviate the issue the packages being downloaded will be the exact version we are expecting (although, it shouldn't be possible to build from a tag and get a different version).